tensorboard/compat/tensorflow_stub/pywrap_tensorflow.py

```diff
@@ -197,19 +197,29 @@ def __init__(
         self.status = status
         self.curr_event = None
         self.file_handle = gfile.GFile(self.filename, "rb")
+        # Maintain a buffer of partially read records, so we can recover from
+        # truncated records upon a retry.
+        self._buffer = b""
+        self._buffer_pos = 0
 
     def GetNext(self):
+        # Each new read should start at the beginning of any partial record.
+        self._buffer_pos = 0
         # Read the header
         self.curr_event = None
-        header_str = self.file_handle.read(8)
-        if len(header_str) != 8:
+        header_str = self._read(8)
+        if not header_str:
             # Hit EOF so raise and exit
             raise errors.OutOfRangeError(None, None, "No more events to read")
+        if len(header_str) < 8:
+            raise self._truncation_error("header")
```

Comment:
AFAIK `read` is allowed to return fewer bytes than were requested, even
if those are available in the stream. (For instance, this can occur when
a signal handler is entered during the read.) From Python’s `read` docs,
it’s not clear to me whether a short result is guaranteed to imply that
EOF is imminent. In practice, it looks like cPython 3.7 and 3.9 both do
exhaust\*. Do you think that it’s worth retrying the `read` until done
or `read` returns `b""`, or is that not necessary?

\* Test program:

<details>
<summary>Source (tested on cPython 3.7.5rc1 and `667b91a5e2` on Debian)</summary>

```python
import os
import signal
import threading
import time


def handle_sigusr1(signalnum, frame):
    print("Got SIGUSR1 (%d)" % signalnum)


def send_delayed_sigusr1(delay):
    time.sleep(delay)
    os.kill(os.getpid(), signal.SIGUSR1)


signal.signal(signal.SIGUSR1, handle_sigusr1)
threading.Thread(target=send_delayed_sigusr1, args=(0.1,)).start()
with open("/dev/zero", "rb") as infile:
    zeros = infile.read(int(1e10))
print(len(zeros))  # seems to always give 1e10 in my testing
```

</details>

Reply:
Thanks for bringing this up and testing it, I hadn't thought about that.  That said, it seems fine to me to not retry the `read` immediately, since A) practically speaking, with this change the higher level code will retry on the next reload cycle anyway and B) under the circumstances you describe, it's not obvious to me that `read` returning zero bytes would actually be a stronger indication of EOF (unless the spec is that it always reads at least 1 byte if available, but might return fewer than available if interrupted by a signal).

Reply:
> unless the spec is that it always reads at least 1 byte if available,
> but might return fewer than available if interrupted by a signal

Yeah, this is basically the spec in C (which is why I ask); Python is
not very clear about what its spec is.

Good point that the higher level code will retry anyway, so the worst
case is that there’ll be a spurious “truncated record” message. Keeping
it as is sounds good to me, then.